### ISSUE TITLE
Fix strange behavior of references in macro definitions and expansions

### ISSUE THREAD

**Issue by sethfowler** [sethfowler/pygmalion#57]
They end up having a source range covering the entire file. Figure out how to avoid that, ideally while retaining the advantage of finding variable references that occur inside macro expansions.

**Comment by sethfowler** [comment]
This going to be ugly with `libclang` as it stands. Here's an example AST dump showing the problem:

```

[FunctionDecl] setToken(statementInfo *const, const tokenType) (c:c.c@20133@F@setToken) @ 511,1 -> 517,2 definition [setToken(statementInfo *const, const tokenType)/c:c.c@20133@F@setToken] reference [setToken(statementInfo *const, const tokenType)/c:c.c@20133@F@setToken]
 [ParmDecl] st (c:c.c@20155@F@setToken@st) @ 511,23 -> 511,46 definition [st/c:c.c@20155@F@setToken@st] reference [st/c:c.c@20155@F@setToken@st]
  [TypeRef] statementInfo () @ 511,23 -> 511,36 definition [statementInfo/c:c.c@5918@T@statementInfo] reference [statementInfo/c:c.c@5918@T@statementInfo]
 [ParmDecl] type (c:c.c@20180@F@setToken@type) @ 511,48 -> 511,68 definition [type/c:c.c@20180@F@setToken@type] reference [type/c:c.c@20180@F@setToken@type]
  [TypeRef] tokenType () @ 511,54 -> 511,63 definition [tokenType/c:c.c@3573@T@tokenType] reference [tokenType/c:c.c@3573@T@tokenType]
 [CompoundStmt]  () @ 512,1 -> 517,2 definition [/] reference [/]
  [DeclStmt]  () @ 513,2 -> 513,19 definition [/] reference [/]
   [VarDecl] token (c:c.c@20205@F@setToken@token) @ 513,2 -> 513,18 definition [token/c:c.c@20205@F@setToken@token] reference [token/c:c.c@20205@F@setToken@token]
    [TypeRef] tokenInfo () @ 513,2 -> 513,11 definition [tokenInfo/c:c.c@5465@T@tokenInfo] reference [tokenInfo/c:c.c@5465@T@tokenInfo]
  [BinaryOperator]  () @ 514,2 -> 514,26 definition [/] reference [/]
   [DeclRefExpr] token () @ 514,2 -> 514,7 definition [token/c:c.c@20205@F@setToken@token] reference [token/c:c.c@20205@F@setToken@token]
   [UnexposedExpr]  () @ 34,29 -> 514,26 definition [/] reference [/]
    [ParenExpr]  () @ 34,29 -> 514,26 definition [/] reference [/]
     [ArraySubscriptExpr]  () @ 34,30 -> 514,26 definition [/] reference [/]
      [UnexposedExpr] token () @ 34,30 -> 514,26 definition [token/c:@S@sStatementInfo@FI@token] reference [token/c:@S@sStatementInfo@FI@token]
       [MemberRefExpr] token () @ 34,30 -> 514,26 definition [token/c:@S@sStatementInfo@FI@token] reference [token/c:@S@sStatementInfo@FI@token]
        [UnexposedExpr]  () @ 34,30 -> 514,26 definition [/] reference [/]
         [ParenExpr]  () @ 34,30 -> 514,26 definition [/] reference [/]
          [DeclRefExpr] st () @ 514,23 -> 514,25 definition [st/c:c.c@20155@F@setToken@st] reference [st/c:c.c@20155@F@setToken@st]
      [CStyleCastExpr]  () @ 34,43 -> 514,26 definition [/] reference [/]
       [UnexposedExpr] tokenIndex () @ 34,49 -> 514,26 definition [tokenIndex/c:@S@sStatementInfo@FI@tokenIndex] reference [tokenIndex/c:@S@sStatementInfo@FI@tokenIndex]
        [MemberRefExpr] tokenIndex () @ 34,49 -> 514,26 definition [tokenIndex/c:@S@sStatementInfo@FI@tokenIndex] reference [tokenIndex/c:@S@sStatementInfo@FI@tokenIndex]
         [UnexposedExpr]  () @ 34,49 -> 514,26 definition [/] reference [/]
          [ParenExpr]  () @ 34,49 -> 514,26 definition [/] reference [/]
           [DeclRefExpr] st () @ 514,23 -> 514,25 definition [st/c:c.c@20155@F@setToken@st] reference [st/c:c.c@20155@F@setToken@st]
```

This nicely shows how things end up screwed up, and how we can still extract the info we need. There's nothing indicating that we're walking a macro expansion except there seems to be an UnexposedExpr at the root. I hope we can rely on that. If we find one of those, we can check if it references a different line than the parent cursor (this should be a cheap filter), and if so, if the start of its source range refers into a macro definition. (we'll have to record these) If so, we know that we're about to walk a macro expansion. Inside, we can charge all references that start in the macro definition to the location of the start of the macro expansion, and all references that don't to the location they give. (Look at the final line - `st` is a parameter to the macro and has a location at the site of the macro's use, not within the macro definition. We definitely want to capture that.)

This will be finicky and awful but it seems doable. I wish that the UnexposedExpr was exposed and the source ranges behaved a little more naturally.

**Comment by sethfowler** [comment]
So I've confirm that this bug is actually _fixed_ in libclang HEAD. That is, the source ranges given above are correct, and there's no jumping backward in the file for the range starts. That means there probably isn't much to do here other than update LLVM. When Pygmalion is released, it will require LLVM 3.3.
